rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,155 @@
+    </Flex>
+  );
+}
+```
+
+This component takes all of its state and business logic from the props passed in.
+
+For use in the actual application, we wrap the component above in a default export which populates the props the
+component needs by using a hook.
+
+```typescript jsx
+export default function Container() {
+  const state = useDesktopSession();
+  return <DesktopSession {...state} />;
+}
+```
+
+
+### Testing
+
+As a result of the above, we end up testing the `DesktopSession` component instead of the actual component (`Container`) that is
+rendered in the UI.
+
+A story may look like this:
+
+```typescript jsx
+export const FetchError = () => (
+  <DesktopSession
+    {...props}
+    fetchAttempt={{ status: 'failed', statusText: 'some fetch  error' }}
+    tdpConnection={{ status: 'success' }}
+    wsConnection={'open'}
+    disconnected={false}
+  />
+);
+```
+
+Which is then tested like so:
+
+```typescript jsx
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the visual part of the UI,
+but skipping all of the business logic.
+
+## Details
+
+### Changing how we test the visual UI
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components, and visually testing the UI. While using snapshot tests of stories is a rough approximation of a visual UI test, they come a lot of baggage: https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e.
+
+It might be worthwhile looking into something like [Chromatic](https://www.chromatic.com/) for visual UI testing.
+
+### Focusing on behavioral tests
+
+We should be testing all the different state possibilities in Jest and React Testing Library, mocking any data or
+network requests to cause the component to render into the state we want. This ensures that the internal logic
+of our components (generally the most critical and difficult-to-get-right aspect of frontend code) gets tested,
+as well as the final UI state.
+
+Tests should check for elements that exist for the specific state being tested, i.e. an error message when the data has
+failed to load. Elements should be clicked on or events triggered in order to change the state of the component.
+
+### Changing how we export components
+
+Default exports should be avoided except when absolutely necessary (for example, with
+`React.Lazy`), as they degrade the developer experience
+for no discernible benefit. For more fully fleshed out reasoning, refer to the articles below:
+- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
+- https://rajeshnaroth.medium.com/avoid-es6-default-exports-a24142978a7a
+- https://blog.piotrnalepa.pl/2020/06/26/default-exports-vs-named-exports/
+- https://ilikekillnerds.com/2019/08/default-exports-bad/
```

Comment:
The references and "for more information see...." should go at the end of the section not within the core content.

Reply:
How comes? They're all links that are only for this section. It seems weird to move them to the bottom, they're linked here to prove the point